packages/odf/components/s3-browser/objects-list/ObjectsList.tsx

```diff
@@ -0,0 +1,330 @@
+import * as React from 'react';
+import {
+  ListObjectsV2CommandOutput,
+  _Object as Content,
+  CommonPrefix,
+} from '@aws-sdk/client-s3';
+import { SelectableTable } from '@odf/shared/table';
+import { useCustomTranslation } from '@odf/shared/useCustomTranslationHook';
+import { useModal } from '@openshift-console/dynamic-plugin-sdk';
+import { LaunchModal } from '@openshift-console/dynamic-plugin-sdk/lib/app/modal-support/ModalProvider';
+import { TFunction } from 'i18next';
+import * as _ from 'lodash-es';
+import { useParams, useSearchParams } from 'react-router-dom-v5-compat';
+import useSWRMutation, { TriggerWithOptionsArgs } from 'swr/mutation';
+import {
+  Button,
+  ButtonVariant,
+  Level,
+  LevelItem,
+  MenuToggle,
+} from '@patternfly/react-core';
+import { AngleLeftIcon, AngleRightIcon } from '@patternfly/react-icons';
+import {
+  ActionsColumn,
+  IAction,
+  CustomActionsToggleProps,
+} from '@patternfly/react-table';
+import { LIST_OBJECTS, DELIMITER, MAX_KEYS, PREFIX } from '../../../constants';
+import { ObjectCrFormat } from '../../../types';
+import { getPath, convertObjectsDataToCrFormat } from '../../../utils';
+import { NoobaaS3Context } from '../noobaa-context';
+import {
+  isRowSelectable,
+  getColumns,
+  TableRow,
+  EmptyPage,
+} from './table-components';
+
+type PaginationProps = {
+  onNext: () => void;
+  onPrevious: () => void;
+  disableNext: boolean;
+  disablePrevious: boolean;
+};
+
+type TableActionsProps = {
+  launcher: LaunchModal;
+  selectedRows: unknown[];
+  loadedWOError: boolean;
+};
+
+type ContinuationTokens = {
+  previous: string[];
+  current: string;
+  next: string;
+};
+
+type Trigger = TriggerWithOptionsArgs<
+  ListObjectsV2CommandOutput,
+  any,
+  string,
+  string
+>;
+
+const continuationTokensSetter = (
+  setContinuationTokens: React.Dispatch<
+    React.SetStateAction<ContinuationTokens>
+  >,
+  response: ListObjectsV2CommandOutput,
+  isNext: boolean,
+  setSelectedRows: React.Dispatch<React.SetStateAction<ObjectCrFormat[]>>
+) => {
+  setContinuationTokens((oldTokens) => {
+    const newTokens = _.cloneDeep(oldTokens);
+    if (isNext) {
+      newTokens.previous.push(newTokens.current);
+      newTokens.current = newTokens.next;
+    } else {
+      newTokens.current = newTokens.previous.pop();
+    }
+    newTokens.next = response.NextContinuationToken;
```

Comment:
```suggestion
    if (isNext) {
      newTokens.previous.push(newTokens.current);
      newTokens.current = newTokens.next;
      newTokens.next = response.NextContinuationToken;
    } else {
      // I am assuming this is for back
       newTokens.next = newTokens.current
      newTokens.current = newTokens.previous.pop();
    }
```

because back noting to do with response.NextContinuationToken I think

Reply:
current conditions in the PR are tested and works well...

Reply:
Ok! it works becuase in esponse.NextContinuationToken and newTokens.current may have same values.

Reply:
But i agree it is safer to use in response.NextContinuationToken for newTokens.next. because,if any object is deleted, then newTokens.current is not the right value to use.